### ISSUE TITLE
`EntityTypeConfiguration<TEntityType>.HasKey` does not allow a nested expression

### ISSUE THREAD

**Issue by AntonC9018** [OData/AspNetCoreOData#828]
**Assemblies affected**

Microsoft.AspNetCore.OData 8.0.12

**Describe the bug**

`EntityTypeConfiguration<TEntityType>.HasKey` does not allow a nested expression.

I'm trying to set up odata with DTOs, mapping to EF Core queries with Automapper.
My DTO, which I'm trying to use as the OData entity, is set up differently than the EF Core entity, it has a nested type with the general information, which has the id on it.
However, HasKey only allows member expressions, so this won't work:
```cs
modelBuilder.EntityType<Dto>()
        .HasKey(a => a.General.Id);
```

**Data Model**

For the purposes of this issue, consider a reduced version:
```cs
public class Dto
{
    public DtoGeneral General { get; set; }
}
public class DtoGeneral
{
    public int Id { get; set; }
}
```

**Additional Notes**

Are nested things even allowed btw?
EF Core has owned types, are all types that aren't entities considered owned in OData?
My model has quite a bunch of nested types, I'd be unfortunate if I had to inline all that.

**Comment by julealgon** [comment]
> My model has quite a bunch of nested types, It'd be unfortunate if I had to inline all that.

Have you thought about just creating a computed redirection `Id` property on the parent that returns/ sets the value in the nested class? 

I find it extremely unlikely that OData will ever support your scenario natively. But maybe I'm wrong.

**Comment by AntonC9018** [comment]
> > My model has quite a bunch of nested types, It'd be unfortunate if I had to inline all that.
> 
> Have you thought about just creating a computed redirection `Id` property on the parent that returns/ sets the value in the nested class?
> 
> I find it extremely unlikely that OData will ever support your scenario natively. But maybe I'm wrong.

Yeah, that's quite an obvious workaround. What about my last question?

**Comment by julealgon** [comment]
> Yeah, that's quite an obvious workaround. What about my last question?

Types that are not entities in OData are considered complex types, not owned. Usually a type is not detected as an entity if it doesn't have an ID field.

There is the concept of an owned entity but that's just an entity that can only be accessed through another and doesn't have it's own entity set. These have to be configured as owned explicitly, either in the EDM, or through attributes.

**Comment by xuzhg** [comment]
Using complex type property to define key is valid in OData spec. 

An example like this:

```xml
<EntityType Name="Category">
  <Key>
    <PropertyRef Name="Info/ID" Alias="EntityInfoID" />
  </Key>
  <Property Name="Info" Type="Sales.EntityInfo" Nullable="false" />
  <Property Name="Name" Type="Edm.String" />
</EntityType>

 

<ComplexType Name="EntityInfo">

  <Property Name="ID" Type="Edm.Int32" Nullable="false" />
  <Property Name="Created" Type="Edm.DateTimeOffset" />
</ComplexType>
```

But, it's not supported ODL. Once ODL supports it, I can make it work at Web API side.

**Comment by habbes** [comment]
@ElizabethOkerio was once looking into adding key and property aliasing support in ODL.
